sui-execution/latest/sui-move-natives/src/crypto/sui_state_proof.rs

```diff
@@ -0,0 +1,287 @@
+        return Ok(NativeResult::err(
+            cost,
+            INVALID_INPUT
+        ));
+    };
+
+
+    let Ok(type_layout) = bcs::from_bytes::<MoveTypeLayout>(&type_layout_bytes) else {
+        return Ok(NativeResult::err(
+            cost,
+            INVALID_INPUT
+        ));
+    };
+
+
+    let Ok(package_id_target) = bcs::from_bytes::<ObjectID>(&package_id_bytes) else {
+        return Ok(NativeResult::err(
+            cost,
+            INVALID_INPUT
+        ));
+    };
+    
+    // Verify the checkpoint summary using the committee
+    let res = summary.verify_with_contents(&committee, Some(&checkpoint_contents));
+    if let Err(err) = res {
+        return Ok(NativeResult::err(cost, INVALID_TX));
+    }
+
+
+    // Ensure the tx is part of the checkpoint
+    let is_valid_checkpoint_tx = checkpoint_contents.iter().any(|&digest| digest == transaction.effects.execution_digests());
+    if !is_valid_checkpoint_tx {
+      return Ok(NativeResult::err(cost, INVALID_TX));
+    };
+
+
+    // transaction.clone().events.unwrap().data.iter().for_each(|event| {
+    //     println!("event {:?}", event);
+    // });
+
+    let tx_events = &transaction.events.as_ref().unwrap().data;
+
+
+    tx_events.iter().for_each(|event| {
+        println!("{:?}", event);
+    });
+    let mut messages = Vec::new();
+    let mut cap_id_final = SuiAddress::ZERO;
+
+    for event in tx_events {
+        if !(event.clone().package_id == package_id_target && event.clone().type_.module.into_string() == "sui_state_proof" && event.clone().type_.name.into_string() == "DWalletNetworkRequest") {
+            continue;
+        }
+
+        let json_val = SuiJsonValue::from_bcs_bytes(Some(&type_layout), &event.contents).unwrap().to_json_value();
+
+
+        let approve_message = match json_val.clone() {
+            JsonValue::Object(map) => {
+                // println!("mapppp{:?}", map);
+                if let Some(msg_value) = map.get("message").and_then(|s| s.as_array()) {
+                    msg_value.to_owned() 
+                }
+                else {
+                    return Ok(NativeResult::err(cost, INVALID_TX));
+                }
+            },
+            _ => return Ok(NativeResult::err(cost, INVALID_TX))
+        };
+
+
+        // TOOD no need to read it several times
+        let cap_id = match json_val.clone() {
+            JsonValue::Object(map) => {
+                if let Some(msg_value) = map.get("cap_id").and_then(|s| s.as_str()) {
+                    msg_value.to_owned() 
+                }
+                else {
+                    return Ok(NativeResult::err(cost, INVALID_TX));
+                }
+    
+            },
+            _ => return Ok(NativeResult::err(cost, INVALID_TX))
+        };
+
+        let cap_id = match SuiAddress::from_str(&cap_id) {
+            Ok(address) => address,
+            Err(_) => return Ok(NativeResult::err(cost, INVALID_TX)),
+        };
+        cap_id_final = cap_id;
```

Comment:
Yeah so I think there's a vulnerability here too. I can create a tx that approves two messages, m1 and m2. m1 I approve for a dWallet1, and m2 I approve for a dWallet2.
Now dWallet2 never did any policy check for m1, and in-fact it can be unallowed by its policy: e.g., you were given access to spend 1 eth in dwallet2. You create a new dwallet1 without any policy (doesn't even have to hold any funds) and approve a 100eth withdrawl from dwallet1 and 0.1 eth from dwallet2 - and because you say the cap id is the last one, and you don't check they are all equal, than the result would be you'll say dwallet2 approve a 100.1 eth withdrawl....
If true this completely breaks the entire security model

Reply:
@alexanderludwig